elasticsearch.md

```diff
@@ -789,36 +787,130 @@ In this example, we want to avoid that users belonging to group "press" can see
+* for responses where fields with values are returned (e.g. Search/Get API) - filter and show only allowed fields
+* make not allowed fields unsearchable - used in QueryDSL requests (e.g. Search/MSearch API) do not have impact on search result.
+
+In other words: FLS protects from usage not allowed fields for user. From user's perspective it seems like such fields are nonexistent. 
+
+**Definition**
+
+Field rule definition consists of two parts:
+
+- non empty list of fields (whitelisted or blacklisted) names. Supports wildcards and user runtime variables.
+- FLS engine definition (global setting, optional)
+
+**Field names**
 
+Fields can be defined using two access modes: whitelist and blacklist.
+ 
 **Whitelist mode**
 
-`fields: ["allowed_fields_prefix_*", "_*", "allowed_field.nested_field.text"]`
+Specifies which fields should be allowed explicitly (other fields from mapping become not allowed implicitly).
+Example:
 
-If the current is a search request, return all matching documents, but deprived of all the fields, except the ones that start with `allowed_fields_prefix_` or with underscore.
+`fields: ["allowed_fields_prefix_*", "_*", "allowed_field.nested_field.text"]`
 
-If you use whitelist mode, remember to allow the mandatory, internally used fields \(the ones that start with underscore, `_*`\).
+Return documents deprived of all the fields, except the ones that:
+ * start with `allowed_fields_prefix_` 
+ * start with underscore
+ * are equal to `allowed_field.nested_field.text`
 
 **Blacklist mode \(recommended\)**
 
+Specifies which fields should not be allowed prefixed with `~` (other fields from mapping become allowed implicitly). Example:
+
 `fields: ["~excluded_fields_prefix_*", "~excluded_field", "~another_excluded_field.nested_field"]`
 
-If the current is a search request, return all matching documents, but deprived of the `excluded_field` and the ones that start with `excluded_fields_prefix_`.
+Return documents but deprived of the fields that:
+  * start with `excluded_fields_prefix_`
+  * are equal to `excluded_field`
+  * are equal to `another_excluded_field.nested_field`
+
+**NB:** You can only provide a full black list or white list. Grey lists \(i.e. `["~a", "b"]`\) are invalid settings and Elasticsearch will refuse to boot up if this condition is detected.
 
-**Example: hide prices from catalogue indices**
+Example: hide prices from catalogue indices
 
 ```text
 - name: "External users - hide prices"
   fields: ["~price"]
   indices: ["catalogue_*"]
 ```
 
+**⚠️IMPORTANT** Any metadata fields e.g. `_id` or `_index` can not be used in fields rule. 
+
+**FLS engine configuration**
+
+Global, optional property `fls_engine` set under `readonlyrest:` configuration section. 
+ 
+FLS engine specifies how ROR handles field level security internally. Previously FLS was based entirely on lucene - that's why ROR needed to be installed on all nodes to make fields rule work properly.
```

Comment:
I'd add link under `Lucene` (capital L) to https://en.wikipedia.org/wiki/Apache_Lucene 
Maybe it's not obvious for everyone